irohad/ametsuchi/impl/postgres_command_executor.cpp

```diff
@@ -528,228 +318,593 @@ namespace iroha {
+                        EXISTS (SELECT value FROM new_src_value
+                                WHERE value >= 0 LIMIT 1) %s
+                  )
+                  ON CONFLICT (account_id, asset_id)
+                  DO UPDATE SET amount = EXCLUDED.amount
+                  RETURNING (1)
+              ),
+              insert_dest AS
+              (
+                  INSERT INTO account_has_asset(account_id, asset_id, amount)
+                  (
+                      SELECT $3, $4, value
+                      FROM new_dest_value
+                      WHERE EXISTS (SELECT * FROM insert_src) AND
+                        EXISTS (SELECT * FROM has_src_account LIMIT 1) AND
+                        EXISTS (SELECT * FROM has_dest_account LIMIT 1) AND
+                        EXISTS (SELECT * FROM has_asset LIMIT 1) AND
+                        EXISTS (SELECT value FROM new_dest_value
+                                WHERE value < 2::decimal ^ (256 - $5)
+                                LIMIT 1) %s
+                  )
+                  ON CONFLICT (account_id, asset_id)
+                  DO UPDATE SET amount = EXCLUDED.amount
+                  RETURNING (1)
+               )
+          SELECT CASE
+              WHEN EXISTS (SELECT * FROM insert_dest LIMIT 1) THEN 0
+              %s
+              WHEN NOT EXISTS (SELECT * FROM has_dest_account LIMIT 1) THEN 2
+              WHEN NOT EXISTS (SELECT * FROM has_src_account LIMIT 1) THEN 3
+              WHEN NOT EXISTS (SELECT * FROM has_asset LIMIT 1) THEN 4
+              WHEN NOT EXISTS (SELECT value FROM new_src_value
+                               WHERE value >= 0 LIMIT 1) THEN 5
+              WHEN NOT EXISTS (SELECT value FROM new_dest_value
+                               WHERE value < 2::decimal ^ (256 - $5)
+                               LIMIT 1) THEN 6
+              ELSE 7
+          END AS result)";
+
+    std::string CommandError::toString() const {
+      return (boost::format("%s: %s") % command_name % error_message).str();
+    }
+
+    PostgresCommandExecutor::PostgresCommandExecutor(soci::session &sql)
+        : sql_(sql), do_validation_(true) {}
+
+    void PostgresCommandExecutor::setCreatorAccountId(
+        const shared_model::interface::types::AccountIdType
+            &creator_account_id) {
+      creator_account_id_ = creator_account_id;
+    }
+
+    void PostgresCommandExecutor::doValidation(bool do_validation) {
+      do_validation_ = do_validation;
+    }
+
+    CommandResult PostgresCommandExecutor::operator()(
+        const shared_model::interface::AddAssetQuantity &command) {
+      auto &account_id = creator_account_id_;
+      auto &asset_id = command.assetId();
+      auto amount = command.amount().toStringRepr();
+      int precision = command.amount().precision();
+
+      auto cmd = boost::format("EXECUTE %1% ('%2%', '%3%', %4%, '%5%')");
+
+      if (do_validation_) {
+        cmd = (cmd % "addAssetQuantityWithValidation");
```

Comment:
Seems this code is duplicated many times. I guess better to refactor it with separate function.